jorldy/run_mode.py

```diff
@@ -295,38 +292,54 @@ def async_distributed_train(config_path, unknown):
             agent.load(config.train.load_path)
 
         save_path = path_queue.get()
-        step, _step, print_stamp, save_stamp = 0, 0, 0, 0
-        while step < config.train.run_step:
-            transitions = []
-            while (_step == 0 or not trans_queue.empty()) and (
-                _step - step < config.train.update_period
-            ):
-                _step, _transitions = trans_queue.get()
-                transitions += _transitions
-            delta_t = _step - step
-            print_stamp += delta_t
-            save_stamp += delta_t
-            step = _step
-            result = agent.process(transitions, step)
+        heap = {
+            "step": 0,
+            "run_step": config.train.run_step,
+            "print_stamp": 0,
+            "save_stamp": 0,
+            "wait_thread": False,
+            "wait_process": True,
+        }
+        step, transitions = 0, []
+        print_signal, save_signal = False, False
+        gath_thread = Thread(
+            target=gather_thread,
+            args=(trans_queue, transitions, heap, "+="),
+        )
+        gath_thread.start()
+        while step < heap["run_step"]:
+            while heap["wait_process"]:
+                time.sleep(0.1)
+            heap["wait_thread"] = True
+            if step != heap["step"]:
+                step = heap["step"]
+                _transitions = transitions.copy()
```

Comment:
In fact, the speed of fetching transitions has been greatly improved.

Test Configuration : Linux Ubuntu, python3.8.8 jorldy0.2.0,  V4 Xlarge, ape-x atrari pong
    "run_step": 10000,
    "print_period": 1000,
    "save_period": 20000,
    "eval_iteration": 5,
    "record": True,
    "record_period": 20000,
    # distributed setting
    "distributed_batch_size": 512,
    "update_period": 100,
    "num_workers": 64,

- Before change: In the pre-change test, there is a bug at the end and it can be inaccurate.
<img width="868" alt="스크린샷 2022-02-28 오전 9 26 31" src="https://user-images.githubusercontent.com/99241809/155906692-5100d840-3baa-40a2-b377-2caf0e5c5631.png">


- after change:
<img width="819" alt="스크린샷 2022-02-28 오전 9 16 41" src="https://user-images.githubusercontent.com/99241809/155906395-b3e03746-e13f-4999-bcd3-7cb5eb44c824.png">